### PR TITLE
pass Consensus::Params& to functions in validation.cpp and make them static

### PR DESCRIPTION
**Pull request by mariodian**
Pass Consensus::Params& to ReceivedBlockTransactions() as requested in #7829

### CODE REVIEW

**Comment by TheBlueMatt** [comment]
utACK. Want to mark ReceivedBlockTransactions static while you're at it?

**Comment by mariodian** [comment]
@TheBlueMatt sure. Should I soft reset the commit and add the new change within a single commit?

**Comment by TheBlueMatt** [comment]
Doesnt matter either way, I think.

**Comment by paveljanik** [comment]
ACK https://github.com/bitcoin/bitcoin/pull/10201/commits/25660e91ef2d399d2db87edb2243122fd4093ede

**Review by dcousens — APPROVED** [review]

**Comment by jtimon** [comment]
ACK 25660e91ef2d399d2db87edb2243122fd4093ede
Added to the list in #7829
If you feel like doing more functions in one go, maybe I can close #10227 (sorry I didn't see this before opening it, you need to "ping @jtimon" to get the promised review faster). Just by equaling it I would be more than happy to close #10227, but it occurs to me that I didn't do an exhaustive search on all the functions in validation.o that could be made static (independently on whether they use CChainParams or Consensus::Params or not). If you do that, I bet not only nobody will complain for increasing the scope of a PR with 3 acks but people will actually praise you (at least I will) for every new function that can be static you find apart from the 3 in #10227.
I think it's also a more interesting and rewarding task than what #7829 proposes.

**Comment by mariodian** [comment]
@jtimon this was the first time I was looking at a C++ code so my understanding of the language is next to zero. However, the similar pattern all the static functions seem to have is that they're only called within the same file. If that's the case (?) then I'll be able to do it (perhaps in a separate PR as you suggested).

**Comment by jtimon** [comment]
Yes, static functions are only called from the same file. So for searching more functions in validation.cpp that could be static, it's more about grep than about C++ (and as said feel free to copy from #10227 ).
Either way I already acked this PR as it is, but would be happy to review more.

**Comment by mariodian** [comment]
ping @jtimon 

Made other methods static (not sure about `FILE* OpenDiskFile()` and `FILE* OpenUndoFile()` though).

Also copied #10227 and added parameter chainparams to `FlushStateToDisk()`.

**Comment by jtimon** [comment]
Yeah, it seems OpenDiskFile and OpenUndoFile are only called from validation.cpp and can be static too.

File: src/validation.h [hunk 1]
```diff
@@ -281,23 +281,6 @@ CAmount GetBlockSubsidy(int nHeight, const Consensus::Params& consensusParams);
 double GuessVerificationProgress(const ChainTxData& data, CBlockIndex* pindex);
 
 /**
- * Prune block and undo files (blk???.dat and undo???.dat) so that the disk space used is less than a user-defined target.
```
Comment (jtimon) [inline src/validation.h]: The documentation you are removing from here, you could move to the cpp with the function definition

File: src/validation.cpp [hunk 2]
```diff
@@ -183,8 +183,8 @@ enum FlushStateMode {
 };
 
 // See definition for documentation
-bool static FlushStateToDisk(CValidationState &state, FlushStateMode mode, int nManualPruneHeight=0);
-void FindFilesToPruneManual(std::set<int>& setFilesToPrune, int nManualPruneHeight);
+bool static FlushStateToDisk(CValidationState &state, FlushStateMode mode, int nManualPruneHeight=0, const CChainParams& chainParams);
```
Comment (jtimon) [inline src/validation.cpp]: Given that nManualPruneHeight is optional, the new CChainParams parameter (which is not) should come before it, for example, CChainParams can be the first parameter.

Comment (mariodian) [inline src/validation.cpp]: Right

**Comment by mariodian** [comment]
@jtimon oh, I "accidentally" soft reset my first commit instead of fixingup the last one so everything is in the single commit now. Sorry, I haven't worked with git for quite a long time.

**Comment by jtimon** [comment]
Mhmm, squashing in one commit is fine, what I mean is you also added a commit that doesn't belong here:
"Merge branch 'master' into consensusparams-receivedblocktransactions" 8b23dfd
Here's what you can do to fix this: ```rebase -i origin/master``` and remove that commit from the list.

**Comment by jtimon** [comment]
Needs rebase

**Comment by mariodian** [comment]
@jtimon I changed the title. 

Before I break anything while rebasing, is the following all that is needed?

```
git checkout consensusparams-receivedblocktransactions
git rebase master
git checkout master
git merge consensusparams-receivedblocktransactions
```

Thank you.

**Comment by jtimon** [comment]
what I would usually do:
```
git checkout consensusparams-receivedblocktransactions
git fetch origin
git rebase -i origin/master
git push jtimon +consensusparams-receivedblocktransactions
```
Instead of jtimon, you need to use something else pointing to you repository
For example, in my bitcoin./git/config I have (among other things):

```
[remote "origin"]
	url = github:bitcoin/bitcoin.git
	fetch = +refs/heads/*:refs/remotes/origin/*
[remote "jtimon"]
	url = github:jtimon/bitcoin.git
	fetch = +refs/heads/*:refs/remotes/jtimon/*
```

Using -i (for interactive) it's not very useful in this case because it is only one commit and the rebase will stop in it for the conflict anyway, but you can try to replace s/pick/edit/ or s/pick/reword/ to experiment with it. You can also use it for squash commits or remove them, I use interactive rebase all the time.

**Comment by mariodian** [comment]
@jtimon I just rebased it. Thanks so much for the help!

**Comment by jtimon** [comment]
You are welcomed, I'm happy to teach and you don't seem afraid to learn.

Travis is already failing: https://travis-ci.org/bitcoin/bitcoin/jobs/226509712#L1865
You should compile locally before pushing (specially before force-pushing [the '+' symbol before your branch name means force push]). Even better if you also run the unittests by doing "make check" (you can use several cores/hyperthreads to build faster with -j4 [example for 4 processes, but depends on you machine]). This PR shouldn't really need it, but you can also run the python tests to be more sure your changes are correct, see https://github.com/bitcoin/bitcoin/tree/master/test (and by the way, writing functional tests is a good way to contribute and learn, plus python is easier to learn than C++).

At least one of your problems seems to be that you need a forward declaration for CheckInputs, but there may be more and you should be able to easily find them and fix them with the help of the compiler.
I'm also jtimon on IRC freenode if you want to ask me questions there too, maybe in #bitcoin-core-dev too if I'm not around. That may be more convenient and also less disruptive to other reviewers.

**Comment by jtimon** [comment]
Although it has been very useful for me to look only at the latest changes you have done to fix the problems separately and it is a nice practice to facilitate reviewers work (ie now I can approve 02cf6e1 independently, without having to re-approve cf8243b which I didn't because it was wrong, but normally it would be just allowing to review the latest changes without re-reviewing anything they have already approved), this PR cannot be merged as it is because not all the commits are correct individually.

Even travis is saying it, but even we didn't had Continuous Integration, we already figured out that cf8243b is incorrect locally. Please, squash again so that every commit is correct individually. People make mistakes, but it is always better for everyone to rewrite git history before it happens (is merged) than afterwards. Since we always maintain at least 2 major versions simultaneously, having a clean, and ideally always-correct history helps maintaining things.

Your only reserve for squashing could be invalidating previous approvals, but note that @paveljanik and @dcousens and me only approved 25660e9 , which this PR no longer contains. That's my fault for asking you to increase the scope of the PR, but I really think just turning functions into static will be more interesting for reviewers than reducing the dependency to the "Params()" function, which not everyone cares about. After you squash, another ACK shouldn't take me much time, this looks good as it is for me.

But I'm also sorry to tell you, but I don't this will be the last time you will have to rebase. Changing the function signature is usually easy to write but often painful to maintain. Maybe not. Some PRs are merged in days, some after months without needing rebase, some need rebase all the time.

One way to solve potential conflicts in advance is by making sure your branch could rebase cleanly (ie without any conflict/modification) on top of all the currently open PRs. That works for small projects, but not for this one, see https://github.com/bitcoin/bitcoin/pulls. You can just wait or have a look or be warned: this conflicts with #10192 (in CheckInputs at least) and probably #10279 too, who knows what else.
You kind of have 3 options:

1) Don't try to solve conflicts in advance, leave things as they are and rebase things as open conflicts get merged.
2)  Avoid the conflicts by reducing the scope of your own PR.
3) Coordinate with the other PR to share preparation commits at the beginning of both or on another PR with only the common parts.

I really look for option 3, but 1 and 2 are usually easier.

**Comment by mariodian** [comment]
@jtimon for now, I'd go for the 1st option which makes it look way easier and less time-consuming. If I contribute in the future, I'd definitely make it less chaotic by reducing the scope of the PR and possibly coordinate with other PRs if necessary.

**Comment by jtimon** [comment]
Needs rebase

**Comment by jtimon** [comment]
needs rebase

File: src/validation.cpp [hunk 3]
```diff
@@ -186,8 +186,11 @@ enum FlushStateMode {
 };
 
 // See definition for documentation
-bool static FlushStateToDisk(CValidationState &state, FlushStateMode mode, int nManualPruneHeight=0);
-void FindFilesToPruneManual(std::set<int>& setFilesToPrune, int nManualPruneHeight);
+bool static FlushStateToDisk(const CChainParams& chainParams, CValidationState &state, FlushStateMode mode, int nManualPruneHeight=0);
```
Comment (laanwj) [inline src/validation.cpp]: `static bool` iso `bool static` would be more consistent with other uses of static in the code

Comment (mariodian) [inline src/validation.cpp]: I agree. Changed it.

**Pull request merged**